h2o-core/src/main/java/water/parser/ParseSetup.java

```diff
@@ -267,11 +273,16 @@ public static ParseSetup guessSetup( Key[] fkeys, ParseSetup userSetup ) {
     //Calc chunk-size
     // FIXME: should be a parser specific - or at least parser should be able to override defaults
     Iced ice = DKV.getGet(fkeys[0]);
+    
     if (ice instanceof Frame && ((Frame) ice).vec(0) instanceof UploadFileVec) {
       t._gblSetup._chunk_size = FileVec.DFLT_CHUNK_SIZE;
     } else {
-      t._gblSetup._chunk_size = FileVec.calcOptimalChunkSize(t._totalParseSize, t._gblSetup._number_columns, t._maxLineLength,
-              Runtime.getRuntime().availableProcessors(), H2O.getCloudSize(), false /*use new heuristic*/, true);
+      int calculatedChunkSize = FileVec.calcOptimalChunkSize(t._totalParseSize, t._gblSetup._number_columns, t._maxLineLength,
+          Runtime.getRuntime().availableProcessors(), H2O.getCloudSize(), false /*use new heuristic*/, true);
+      int maxExpectedHeaderSize = Math.max(10000, (int)t._maxLineLength * 300); // to make sure we dont lose bits and get a negative
+      int acceptableHeaderSize = Math.min(500000, maxExpectedHeaderSize); //can't expect half a meg of headers
+      int finalChunkSize = Math.max(calculatedChunkSize, acceptableHeaderSize);
```

Comment:
The current algorithm requires that the whole header fits in one chunk. That was the problem this change fixes. Later I'll go ahead and use a better algorithm; there are pieces of code that were probably developed for that purpose. Have to work on it.
This specific change kind of makes sure that a reasonable-sized header will fit into one chunk. Making small chunks makes no sense when reading files, it's 2017.
Still, since maxLineLength is, at times, calculated incorrectly (I have some examples), we have to be careful.